### ISSUE TITLE
Access to unitialized var

### ISSUE THREAD

**Issue by ncannasse** [HaxeFoundation/haxe#4849]
The following abstract (found in `Issue4327.hx`)

```haxe
@:callable
abstract Example(Void->String) {
	public function new() {
		this = fun;
	}

	function fun() {
		return "foo";
	}
}
```

Will produce the following AST:

```
	static public _new(method) : Void -> Example
	 = [Function:Void -> Example]
		[Block:Dynamic]
			[Var this1(1864):Void -> String]
			[Var _e(2680):Example] [Local this1(1864):Void -> String]
```

As you can notice, here we initialize `_e` with `this1` which is not initialized.

**Comment by Simn** [comment]
Hmm, not sure why this doesn't cause a "Missing this = value" from `check_local_vars_init`.

**Comment by ncannasse** [comment]
Most likely this comes from some transform that runs after it?

**Comment by Simn** [comment]
This is a more general problem related to uninitialized variables in closures:

```haxe
class Main  {
    static function main() {
		var x;
		function f() {
			trace(x);
		}
	}
}
```

```
HL Interp error Can't set enumfield(dyn) with undef
Main.hx:4: Called from fun(144)@1
<startup>:1: Called from fun(396)@804
```

**Comment by ncannasse** [comment]
Ah, it is right to detect such cases given that x might be actually initialized after the method is declared. Let's keep it as an HL-specific fix for now.

**Comment by ncannasse** [comment]
That's pretty hard to know which variables should be force-initialized actually. Maybe that's a job for Haxe compiler and not code generator... I don't think we want to error in these cases.

**Comment by Simn** [comment]
Ideally function bodies should be checked upon the first usage of their variables. The case here is quite artificial because the local function is never used, perhaps this is more realistic:

```haxe
class Main  {
	static function main() {
		var x;
		function f() {
			trace(x);
		}
		f();
		x = 1;
	}
}
```

If we can detect this use-before-def case we can give an error here. Conversely, if the assignment is made before the call then everything is good. This is not only about calls to `f` but also using it as a value in other places because we can't really tell who might call it.

**Comment by ncannasse** [comment]
I agree that's maybe the best way to deal with it. This might cause a few errors when we enable it, but it should not be widespread.
